### DISCUSSION TITLE
How to know if PDF can have watermark (under text) or just a stamp (on top)?

### DISCUSSION THREAD

**Discussion by zyzlik** [py-pdf/pypdf#1197] [Q&A] [answered]
I have different PDF files: some are digitally generated and some are scans / images. I have to put watermark on them and preferable under the text. I used `page.extract_text()` to determine if PDF is digitally-generated and if so, I added watermark under the text, otherwise on top. But for some digitally generated PDFs (e.g. Amazon invoice) I am able to extract text, but watermark is not visible behind the text, while for some it works normally.

Is there a better way to determine if watermark can be behind text?

**Comment by MartinThoma** [comment, marked as answer]
The issue you encounter is that the watermark goes behind an image. You could check if the page contains any image.

Something like this:

```python
def contains_image(page: PageObject) -> bool:
    page_resources = page["/Resources"]

    x_object = page_resources.get("/XObject", {})

    for obj in x_object:
        obj_ = x_object[obj]
        if obj_["/Subtype"] == "/Image":
            return True
    return False
```

You could enhance this with the image dimensions. Or you could look at metadata (the generator). If you see Epson / Canon or similar in there, there is a good chance that it's a scanner.

**Reply by zyzlik** [reply]
Oh, that should work, thanks!

**Reply by zyzlik** [reply]
But! I tested this on the pdf with text, where underlying watermark works and it also contains images (logo), but watermark is displayed properly behind the text.
And for pdf where watermark doesn't work contains image 113x23 while page itself is 612x792 🤷‍♀️

**Reply by MartinThoma** [reply]
Yes, it's not a bullet-proof method. You could use pymupdf (Fitz) to render the complete page before and after as an image. If the image changed, you know that at least a part of the watermark is visible

**Reply by MartinThoma** [reply]
You could also look at the size / position of the image
